### PR TITLE
Truncate range tombstones from sstables

### PR DESCRIPTION
**Pull request by petermattis**
[This is a backport to crl-release-5.13 of an upstream PR. I'll do my best to keep the implementations in sync as comments come in on both PRs.]

This PR provides little benefit by itself, but it allows us to relax a compaction heuristic that expands a compaction to include all of the sstables that are overlapped by a range tombstone which can result in excessively large compactions. That change will be performed in a follow-on PR.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3)

### CODE REVIEW

**Review by benesch — APPROVED** [review]
:lgtm: modulo the remaining TODO

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-:-LH2xXrmB43Bgi_u8usg:btz73mi)* status: all files reviewed, 1 unresolved discussion (waiting on @petermattis and @bdarnell)

___
*[db/range_del_aggregator.cc, line 217 at r1](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-LH2xSQL4fxk0lYsxblk:-LH2xSQMBrQ5_Pt23a7M:b-qasc0i) ([raw file](https://github.com/cockroachdb/rocksdb/blob/e8bf45470a4f921228b1aeaa67473298dec3fe46/db/range_del_aggregator.cc#L217)):*
> ```cpp
>     // TODO(peter): What is the lifetime smallest and largest? They
>     // are being taken from FileMetadata::{smallest,largest}. Will
>     // they live at least as long as the RangeDelAggregator?
> ```

We should resolve this TODO before merging.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3)* status: 9 of 10 files reviewed, 1 unresolved discussion (waiting on @benesch, @petermattis, and @bdarnell)

___
*[db/range_del_aggregator.cc, line 217 at r1](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-LH2xSQL4fxk0lYsxblk:-LH3GOjs8goa7NrhcNQn:b-1006dg) ([raw file](https://github.com/cockroachdb/rocksdb/blob/e8bf45470a4f921228b1aeaa67473298dec3fe46/db/range_del_aggregator.cc#L217)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We should resolve this TODO before merging.
</blockquote></details>

I tracked down the lifetime of `FileMetaData::{smallest,largest}`. They are tied to `VersionStorageInfo` which itself is tied to `Version`. An iterator pins the `Version` it is associated with in memory via a reference to a `SuperVersion`. I believe there is nothing to do here.

**Review by benesch — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bdarnell)

**Review by tbg — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-:-LH4LE-N3i03wZzr6A4t:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bdarnell)

**Review by bdarnell — APPROVED** [review]
> Reviewed 9 of 10 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3)* status: all files reviewed, 1 unresolved discussion (waiting on @petermattis)

___
*[db/range_del_aggregator.h, line 107 at r2](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-LH4XGYl1mL35AjorTaR:-LH4XGYl1mL35AjorTaS:bts4gjz) ([raw file](https://github.com/cockroachdb/rocksdb/blob/6c31640bf5ef446be284ef649522e6a6a66b825a/db/range_del_aggregator.h#L107)):*
> ```objc
> 
>   // Adds tombstones to the tombstone aggregation structure maintained
>   // by this object. Tombstones are truncated to smallest and
> ```

This could probably use a comment about inclusive vs exclusive end keys.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/rocksdb/3)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @bdarnell and @benesch)

___
*[db/range_del_aggregator.h, line 107 at r2](https://reviewable.io/reviews/cockroachdb/rocksdb/3#-LH4XGYl1mL35AjorTaR:-LH4a7n99difehH-Yeq7:b-896fix) ([raw file](https://github.com/cockroachdb/rocksdb/blob/6c31640bf5ef446be284ef649522e6a6a66b825a/db/range_del_aggregator.h#L107)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This could probably use a comment about inclusive vs exclusive end keys. 
</blockquote></details>

Done.

**Pull request merged**
